### PR TITLE
docs for single-host devfile endpoins exposure on subdomains

### PR DESCRIPTION
**Pull request by sparkoo**
Signed-off-by: Michal Vala <mvala@redhat.com>

> Read our [Contribution guide](https://github.com/eclipse/che-docs/blob/master/CONTRIBUTING.adoc) before submitting a PR.

### What does this PR do?
Document new single-host configuration for exposing devfile endpoints on subdomains.
Che PR: https://github.com/eclipse/che/pull/17928

### What issues does this PR fix or reference?
https://github.com/eclipse/che/issues/17840

### Specify the version of the product this PR applies to.
7.20

### PR Checklist

As the author of this Pull Request I made sure that:

- [x] `vale` has been run successfully against the PR branch
- [x] Link checker has been run successfully against the PR branch
- [x] Documentation describes a scenario that is already covered by QE tests, otherwise an issue has been created and acknowledged by Che QE team
- [ ] Changed article references are updated where they are used (or a redirect has been set up on the docs side):
    - [ ] Dashboard [branding.constant.ts](https://github.com/eclipse/che-dashboard/blob/master/src/components/branding/branding.constant.ts) + [product.json](https://github.com/eclipse/che-dashboard/blob/master/src/assets/branding/product.json)
    - [ ] Chectl [constants.ts](https://github.com/che-incubator/chectl/blob/master/src/constants.ts)

### CODE REVIEW

**Review by skabashnyuk — APPROVED** [review]

**Comment by sparkoo** [comment]
this is the issue that was created during initial single-host testing https://github.com/eclipse/che/issues/17783. @artaleks9 is handling this for QE.

File: modules/installation-guide/partials/con_workspace-exposure-strategies.adoc [hunk 1]
```diff
@@ -34,13 +33,26 @@ When any of those single-host types is used, all workspaces are deployed to sub-
 
 This is convenient for TLS-secured {prod-short} servers because it is sufficient to have a single certificate for the {prod-short} server, which will cover all the workspace component deployments as well.
 
-This strategy limits the exposed components and user applications. Any absolute URL generated on the server side that points back to the server does not work. This is because the server is hidden behind a path-rewriting Ingress that hides the workspace and the component-specific URL prefix from the server.
+There are two types of endpoints, that are defined in the devfiles, exposures in single-host strategy. To configured it, use `++CHE_INFRA_KUBERNETES_SINGLEHOST_WORKSPACE_DEVFILE__ENDPOINT__EXPOSURE++`. It is effective only in single-host server strategy..
```
Comment (metlos) [inline modules/installation-guide/partials/con_workspace-exposure-strategies.adoc]: I'm not sure if English uses the "implanted" sentences the same way as Czech. For me, the English version is hard to follow unless I translate it into Czech at which point it suddenly makes much more sense :)

What about rewording it as:

There are 2 ways of exposing the endpoints specified in the devfile. These can be configured using the `CHE_INFRA_KUBERNETES_SINGLEHOST_WORKSPACE_DEVFILE__ENDPOINT__EXPOSURE` environment variable of the `{prod-short}`. This environment variable is only effective with the single-host server strategy and is applicable to all workspaces of all users.

File: modules/installation-guide/partials/con_workspace-exposure-strategies.adoc [hunk 2]
```diff
@@ -34,13 +33,26 @@ When any of those single-host types is used, all workspaces are deployed to sub-
...
+
+=== devfile endpoints: `single-host`
+
+`++CHE_INFRA_KUBERNETES_SINGLEHOST_WORKSPACE_DEVFILE__ENDPOINT__EXPOSURE: 'single-host'++`
+
+This single-host configuration exposes the endpoints on subpaths, e.g.: `++https://<che-host>/serverihzmuqqc/go-cli-server-8080++`. This limits the exposed components and user applications. Any absolute URL generated on the server side that points back to the server does not work. This is because the server is hidden behind a path-rewriting Ingress that hides the workspace and the component-specific URL prefix from the server.
```
Comment (metlos) [inline modules/installation-guide/partials/con_workspace-exposure-strategies.adoc]: I'd replace "Ingress" with "reverse proxy", because that's what makes the endpoint not work.

Also, "that hides the workspace and the component-specific URL prefix from the server." - the "server" here is confusing, because we're talking about components and user applications in the prior sentence. I would probably say "that hides the unique URL path prefix from the component or user application."

File: modules/installation-guide/partials/con_workspace-exposure-strategies.adoc [hunk 3]
```diff
@@ -34,13 +33,26 @@ When any of those single-host types is used, all workspaces are deployed to sub-
...
 
 For example, when the user accesses the hypothetical `[{prod-url}/component-prefix-djh3d/app/index.php]` URL, the application sees the request coming to `++https://internal-host/app/index.php++`. If the application used the host in the URL that it generates in its UI, it would not work because the internal host is different from the externally visible host. However, if the application used an absolute path as the URL (for the example above, this would be `/app/index.php`), such URL would still not work. This is because on the outside, such URL does not point to the application, because it is missing the component-specific prefix.
 
 Therefore, only applications that use relative URLs in their UI work with the single-host workspace exposure strategy.
 
+=== devfile endpoints: `multi-host`
+`++CHE_INFRA_KUBERNETES_SINGLEHOST_WORKSPACE_DEVFILE__ENDPOINT__EXPOSURE: 'multi-host'++`
 
+This single-host configuration exposes the endpoints on subdomains, e.g.: `++http://serverihzmuqqc-go-cli-server-8080.<che-host>++`. These endpoints will be exposed on unsecured HTTP port. We're using dedicated Ingress or Route for such endpoints, even with `gateway` single-host setup.
```
Comment (metlos) [inline modules/installation-guide/partials/con_workspace-exposure-strategies.adoc]: This single-host configuration exposes the endpoints on subdomains, e.g.: `++http://serverihzmuqqc-go-cli-server-8080.<che-host>++`. These endpoints are be exposed on an unsecured HTTP port. A dedicated Ingress or Route is used for such endpoints, even with `gateway` single-host setup.

File: modules/installation-guide/partials/con_workspace-exposure-strategies.adoc [hunk 4]
```diff
@@ -34,13 +33,26 @@ When any of those single-host types is used, all workspaces are deployed to sub-
...
+
+This configuration limits the usability of previews shown directly in the editor page when {prod-short} is configured with TLS. Because `https` pages allows communication only with secured endpoints, users must open their application previews on another tab of the browser.
```
Comment (metlos) [inline modules/installation-guide/partials/con_workspace-exposure-strategies.adoc]: "in another browser tab".

File: modules/installation-guide/partials/con_workspace-exposure-strategies.adoc (hunk 1 above)
Comment (sparkoo) [reply modules/installation-guide/partials/con_workspace-exposure-strategies.adoc]: yes, fixed

File: modules/installation-guide/partials/con_workspace-exposure-strategies.adoc (hunk 3 above)
Comment (sparkoo) [reply modules/installation-guide/partials/con_workspace-exposure-strategies.adoc]: fixed

File: modules/installation-guide/partials/con_workspace-exposure-strategies.adoc (hunk 4 above)
Comment (sparkoo) [reply modules/installation-guide/partials/con_workspace-exposure-strategies.adoc]: fixed

File: modules/installation-guide/partials/con_workspace-exposure-strategies.adoc (hunk 2 above)
Comment (sparkoo) [reply modules/installation-guide/partials/con_workspace-exposure-strategies.adoc]: this is just copy-paste of what was already there. good time to refresh it. fixed.

File: modules/installation-guide/partials/con_workspace-exposure-strategies.adoc [hunk 5]
```diff
@@ -34,13 +33,26 @@ When any of those single-host types is used, all workspaces are deployed to sub-
...
+There are 2 ways of exposing the endpoints specified in the devfile. These can be configured using the `++CHE_INFRA_KUBERNETES_SINGLEHOST_WORKSPACE_DEVFILE__ENDPOINT__EXPOSURE++` environment variable of the {prod-short}. This environment variable is only effective with the single-host server strategy and is applicable to all workspaces of all users.
+
+=== devfile endpoints: `single-host`
+
+`++CHE_INFRA_KUBERNETES_SINGLEHOST_WORKSPACE_DEVFILE__ENDPOINT__EXPOSURE: 'single-host'++`
+
+This single-host configuration exposes the endpoints on subpaths, e.g.: `++https://<che-host>/serverihzmuqqc/go-cli-server-8080++`. This limits the exposed components and user applications. Any absolute URL generated on the server side that points back to the server does not work. This is because the server is hidden behind a path-rewriting reverse proxy that hides the unique URL path prefix from the component or user application.
```
Comment (yhontyk) [inline modules/installation-guide/partials/con_workspace-exposure-strategies.adoc]:
```suggestion
This single-host configuration exposes the endpoints on subpaths, for example: `++https://<che-host>/serverihzmuqqc/go-cli-server-8080++`. This limits the exposed components and user applications. Any absolute URL generated on the server side that points back to the server does not work. This is because the server is hidden behind a path-rewriting reverse proxy that hides the unique URL path prefix from the component or user application.
```

File: modules/installation-guide/partials/con_workspace-exposure-strategies.adoc [hunk 6]
```diff
@@ -34,13 +33,26 @@ When any of those single-host types is used, all workspaces are deployed to sub-
...
 
 For example, when the user accesses the hypothetical `[{prod-url}/component-prefix-djh3d/app/index.php]` URL, the application sees the request coming to `++https://internal-host/app/index.php++`. If the application used the host in the URL that it generates in its UI, it would not work because the internal host is different from the externally visible host. However, if the application used an absolute path as the URL (for the example above, this would be `/app/index.php`), such URL would still not work. This is because on the outside, such URL does not point to the application, because it is missing the component-specific prefix.
 
 Therefore, only applications that use relative URLs in their UI work with the single-host workspace exposure strategy.
 
+=== devfile endpoints: `multi-host`
+`++CHE_INFRA_KUBERNETES_SINGLEHOST_WORKSPACE_DEVFILE__ENDPOINT__EXPOSURE: 'multi-host'++`
 
+This single-host configuration exposes the endpoints on subdomains, e.g.: `++http://serverihzmuqqc-go-cli-server-8080.<che-host>++`. These endpoints are exposed on an unsecured HTTP port. A dedicated Ingress or Route is used for such endpoints, even with `gateway` single-host setup.
```
Comment (yhontyk) [inline modules/installation-guide/partials/con_workspace-exposure-strategies.adoc]:
```suggestion
This single-host configuration exposes the endpoints on subdomains, for example: `++http://serverihzmuqqc-go-cli-server-8080.<che-host>++`. These endpoints are exposed on an unsecured HTTP port. A dedicated Ingress or Route is used for such endpoints, even with `gateway` single-host setup.
```

File: modules/installation-guide/partials/con_workspace-exposure-strategies.adoc [hunk 7]
```diff
@@ -34,13 +33,26 @@ When any of those single-host types is used, all workspaces are deployed to sub-
...
+
+This configuration limits the usability of previews shown directly in the editor page when {prod-short} is configured with TLS. Because `https` pages allows communication only with secured endpoints, users must open their application previews in another browser tab.
```
Comment (yhontyk) [inline modules/installation-guide/partials/con_workspace-exposure-strategies.adoc]:
```suggestion
This configuration limits the usability of previews shown directly in the editor page when {prod-short} is configured with TLS. Since `https` pages allows communication only with secured endpoints, users must open their application previews in another browser tab.
```

Comment (yhontyk) [inline modules/installation-guide/partials/con_workspace-exposure-strategies.adoc]:
```suggestion
This configuration limits the usability of previews shown directly in the editor page when {prod-short} is configured with TLS. Because `https` pages allow communication only with secured endpoints, users must open their application previews in another browser tab.
```

**Review by yhontyk — APPROVED** [review]
Some small language suggestions, otherwise LGTM, approving.

File: modules/installation-guide/partials/con_workspace-exposure-strategies.adoc [hunk 8]
```diff
@@ -34,13 +33,26 @@ When any of those single-host types is used, all workspaces are deployed to sub-
...
 
 Therefore, only applications that use relative URLs in their UI work with the single-host workspace exposure strategy.
 
+=== devfile endpoints: `multi-host`
+`++CHE_INFRA_KUBERNETES_SINGLEHOST_WORKSPACE_DEVFILE__ENDPOINT__EXPOSURE: 'multi-host'++`
 
+This single-host configuration exposes the endpoints on subdomains, for example: `++http://serverihzmuqqc-go-cli-server-8080.<che-host>++`. These endpoints are exposed on an unsecured HTTP port. A dedicated Ingress or Route is used for such endpoints, even with `gateway` single-host setup.
+
+This configuration limits the usability of previews shown directly in the editor page when {prod-short} is configured with TLS. Because `https` pages allow communication only with secured endpoints, users must open their application previews in another browser tab.
```
Comment (sparkoo) [inline modules/installation-guide/partials/con_workspace-exposure-strategies.adoc]:
```suggestion
This configuration limits the usability of previews shown directly in the editor page when {prod-short} is configured with TLS. Since `https` pages allow communication only with secured endpoints, users must open their application previews in another browser tab.
```
